app/models/group_tag_association.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+class GroupTagAssociation < ActiveRecord::Base
+  belongs_to :group
+  belongs_to :tag
+
+  def self.batch_set(group, tag_names)
+    tag_names ||= []
+    changed = false
+
+    records = self.where(group: group)
+    old_ids = records.pluck(:tag_id)
+
+    tag_ids = tag_names.empty? ? [] : Tag.where_name(tag_names).pluck(:id)
+
+    Tag.where_name(tag_names).joins(:target_tag).each do |tag|
+      tag_ids[tag_ids.index(tag.id)] = tag.target_tag_id
```

Comment:
It would be simpler and faster if you used a `Set.new` for `tag_ids`. You wouldn't have to call `uniq!` below and replacing it would not involve a linear search for the `id` because you could do this:

tag_ids.delete(tag.id)
tag_ids << tag.target_tag_id